### DISCUSSION TITLE
Question - Buffer size too small?

### DISCUSSION THREAD

**Discussion by jjwattzz** [strongswan/strongswan#1517] [General]
Hey, thanks for reading friend. 

I am troubleshooting a very curious connection issue that only occurs on our EC2 AWS hosted instances. For some reason, everything being the same on instances that work just fine, we get the following error when attempting an IPSEC IKEv2 connection via the internet interface. It seems like strongswan can't process the recieved TLS cert because it is too large, and the buffer isn't big enough? Is there any way to increase the buffer size to test this? I have been reading through documentation but haven't found anything. 

![image](https://user-images.githubusercontent.com/106835058/213528531-7c911967-d9de-43e1-a399-a587bf8d5caa.png)

**Comment by strongX509** [comment]
Hi,
I'm rather worried about the previous error message "ASN.1 object smaller than 2 octets" which indicates that what you are trying to load is not an X.509 certificate.

**Reply by jjwattzz** [reply]
Really? Hmmm... see the weird thing is we have tested out this cert and connection with another one of our instances, built the same way, outside of our DMZ. This instance is within the DMZ and directly accessible to the interwebs. I wonder of AWS is modifying the identity cert in some way?

Our charon logs do, however, show that it is indeed a certificate. I will post them.

**Comment by jjwattzz** [comment]
10[TLS] <CEBWGCM|1> received TLS server certificate 'C=CA, O=Carillon Information Security Inc., OU=Devices, OU=Aircraft Operator Ground Stations, OU=CEB, CN=WGL023 - ID - ID'
10[ASN] <CEBWGCM|1> L0 - x509: ASN.1 object smaller than 2 octets
10[LIB] <CEBWGCM|1> OpenSSL X.509 parsing failed
10[LIB] <CEBWGCM|1> building CRED_CERTIFICATE - X509 failed, tried 4 builders
10[TLS] <CEBWGCM|1> parsing TLS certificate failed, skipped
10[ASN] <CEBWGCM|1> L0 - x509: ASN.1 object smaller than 2 octets

**Comment by jjwattzz** [comment]
Further up in the logs is verification that the thing being passed to the function is indeed a certificate, I left that out due to how long it was, but here are the text files of the certificate being passed to the function. You can see that in the working case (135) it is a small certificate. In the AWS case it is a certificate, but it has been parsed and separated into many files that all get sent individually, not sure. And its a LOT bigger. Thanks for the advice in checking the cert being sent, now to find out WHY it's being sent in this format. Would strongswan have anything to do with that? Can it ask for a particular format?

**Reply by tobiasbrunner** [reply]
Please post the complete and unedited logs.

**Reply by jjwattzz** [reply]
Not a problem and thank you for the feedback, we have been banging our heads on this one for weeks. 

The 'IKE_AUTH response 5' line is where the error occurs in AWS log. Log 135 is a successful attempt with a similar (99% the same) AWS EC2 instance setup except it is behind a transit gateway, using the same identity certificate. The only difference is all traffic must go through a firewall and TGW to get to the 135 instance, whereas the AWS's only line of defense is the windows firewall.

**Reply by tobiasbrunner** [reply]
Something is seriously wrong here. In the successful case we see this:
```
15[TLS] <WGL135|1> buffering 358 bytes, 4224 bytes of 4224 byte TLS record received
15[TLS] <WGL135|1> processing buffered TLS Handshake record (4219 bytes)
15[TLS] <WGL135|1> received TLS ServerHello handshake (70 bytes)
15[TLS] <WGL135|1> negotiated TLS 1.2 using suite TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
15[TLS] <WGL135|1> received TLS Certificate handshake (3774 bytes)
15[ASN] <WGL135|1> L0 - x509:
15[ASN] <WGL135|1> => 1663 bytes @ 0x7fab34003d50
...
15[ASN] <WGL135|1> L1 - signatureValue:
15[ASN] <WGL135|1> => 513 bytes @ 0x7fab340041ce
15[ASN] <WGL135|1>    0: 00 5E 74 A3 B8 8B 48 00 56 01 43 BB B6 7C B1 3F  .^t...H.V.C..|.?
...
15[ASN] <WGL135|1>  240: 54 55 3B CC D5 3A B4 12 42 61 18 E5 50 59 1C D4  TU;..:..Ba..PY..
15[ASN] <WGL135|1>  256: FE 3E 18 3C 6D 83 69 0D 50 72 C1 45 50 2D 3F 34  .>.<m.i.Pr.EP-?4
...
15[ASN] <WGL135|1>  496: AC DF D3 AC F1 59 A9 BD 27 47 42 70 25 3E 2C 6B  .....Y..'GBp%>,k
15[ASN] <WGL135|1>  512: 32                                               2
15[TLS] <WGL135|1> received TLS server certificate 'C=CA, O=Carillon Information Security Inc., OU=TEST, OU=Devices, OU=Aircraft Operator Ground Stations, OU=Teledyne Controls, CN=ELS-VPAPP-WGL15 - ID'
15[ASN] <WGL135|1> L0 - x509:
15[ASN] <WGL135|1> => 2102 bytes @ 0x7fab34005660
15[ASN] <WGL135|1>    0: 30 82 08 32 30 82 06 1A A0 03 02 01 02 02 03 14  0..20...........
```
While the corresponding section in the log of the failing case looks like this:
```
10[TLS] <WGL135|1> buffering 184 bytes, 4256 bytes of 4256 byte TLS record received
10[TLS] <WGL135|1> processing buffered TLS Handshake record (4251 bytes)
10[TLS] <WGL135|1> received TLS ServerHello handshake (70 bytes)
10[TLS] <WGL135|1> negotiated TLS 1.2 using suite TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
10[TLS] <WGL135|1> received TLS Certificate handshake (3774 bytes)
10[ASN] <WGL135|1> L0 - x509:
10[ASN] <WGL135|1> => 1663 bytes @ 0x7f2a98003d10
...
10[ASN] <WGL135|1> L1 - signatureValue:
10[ASN] <WGL135|1> => 513 bytes @ 0x7f2a9800418e
10[ASN] <WGL135|1>    0: 00 5E 74 A3 B8 8B 48 00 56 01 43 BB B6 7C B1 3F  .^t...H.V.C..|.?
...
10[ASN] <WGL135|1>  240: 54 55 3B CC D5 3A B4 12 42 61 18 00 00 00 00 00  TU;..:..Ba......
10[ASN] <WGL135|1>  256: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
...
10[ASN] <WGL135|1>  496: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
10[ASN] <WGL135|1>  512: 00                                               .
10[TLS] <WGL135|1> received TLS server certificate 'C=CA, O=Carillon Information Security Inc., OU=TEST, OU=Devices, OU=Aircraft Operator Ground Stations, OU=Teledyne Controls, CN=ELS-VPAPP-WGL15 - ID'
10[ASN] <WGL135|1> L0 - x509:  ASN.1 object smaller than 2 octets
10[LIB] <WGL135|1> OpenSSL X.509 parsing failed
10[LIB] <WGL135|1> building CRED_CERTIFICATE - X509 failed, tried 4 builders
10[TLS] <WGL135|1> parsing TLS certificate failed, skipped
```
So for some reason there are only zeros available after a certain point. You see the decryption of that message starting at line 68154 in the log (the plain text data ends with 544 bytes of zeroes instead of the actual certificate data etc.). And while the end-entity certificate is parsed, it definitely wouldn't be usable due to the defective signature value. And trying to parse an intermediate CA certificate afterwards from only zero bytes will obviously fail (and it does so repeatedly).

Since the padding and its length is correctly decrypted (see line 68423), I don't think this is an issue on the strongSwan side during decryption. So it's probably a problem on the sender side. But what exactly could fail there in this way, no idea. It's also weird that the next IKE message that's received (its decryption starts at line 69096) and contains the rest of that 4256 byte TLS record does not contain zeros only.

**Reply by jjwattzz** [reply]
As we suspected, it seems that AWS is modifying the end-identity certificate before sending it, somehow. We have some more tests lined up and I will report back the solution when we get it. Greatly appreciate you guys, strongswan is awesome!

**Reply by tobiasbrunner** [reply]
Not only the end-entity certificate. Zeroed out are complete sections of the TLS exchange: the end of the end-entity certificate as well as the beginning and again the end of the intermediate CA certificate. Maybe it's a problem with (EAP-)TLS fragmentation. Because the zeroes are in two larger EAP-TLS payloads, both 2044 bytes and in both cases the zeroes start at byte 1500 (MTU?). Unfortunately, only the sender can control the fragment size.
